pkg/loadbalancer/smooth_weighted_round_robin_test.go

```diff
@@ -0,0 +1,310 @@
+		if err != nil {
+			return 10.0
+		}
+		f := float64(rInt.Int64()) / (1 << 63)
+		return minWeight + f*(maxWeight-minWeight)
+	}
+
+	addServer := func(index int, lb *smoothWeightedRR) server {
+		server := servers[index]
+		err := lb.Add(server.name, server.weight)
+		if err != nil {
+			Fail(err.Error())
+		}
+		return server
+	}
+
+	addAllServers := func(lb *smoothWeightedRR) {
+		for i := 0; i < len(servers); i++ {
+			_ = addServer(i, lb)
+		}
+	}
+
+	getTotalServesWeight := func(servers []server) float64 {
+		var weight float64
+		for _, v := range servers {
+			weight += v.weight
+		}
+		return weight
+	}
+
+	// Validations
+	validateServerAdded := func(index int, lb *smoothWeightedRR) {
+		server := servers[index]
+		serversMap := lb.itemsMap
+		Expect(serversMap[server.name].weight).To(Equal(server.weight))
+	}
+
+	validateAllServersAdded := func(lb *smoothWeightedRR) {
+		for i := 0; i < len(servers); i++ {
+			validateServerAdded(i, lb)
+		}
+	}
+
+	validateEmptyLBState := func(lb *smoothWeightedRR) {
+		Expect(lb.ItemsCount()).To(Equal((0)))
+		for i := 0; i < 100; i++ {
+			Expect(lb.Next()).To(BeNil())
+		}
+	}
+
+	validateLoabBalancing := func(rounds int, servers []server, lb *smoothWeightedRR) {
+		totalServersWeight := getTotalServesWeight(servers)
+
+		results := make(map[string]float64)
+
+		for i := 0; i < rounds; i++ {
+			s := lb.Next().(string)
+			results[s]++
+		}
+
+		for i := 0; i < len(servers); i++ {
+			s := servers[i]
+			expectedWeight := float64(rounds) * s.weight / totalServersWeight
+			// Expected weight should be +-1 from the weight counted
+			Expect(results[s.name]).Should(Or(Equal(math.Ceil(expectedWeight)), Equal(math.Floor(expectedWeight))))
+		}
+	}
+
+	BeforeEach(func() {
+		lb = NewSmoothWeightedRR().(*smoothWeightedRR)
+		smoothTestingServers = []server{
+			{name: "server1", weight: 5},
+			{name: "server2", weight: 1},
+			{name: "server3", weight: 1},
+		}
+		rand.Seed(time.Now().UnixNano())
+		servers = []server{
+			{name: "server1", weight: randFloat()},
+			{name: "server2", weight: randFloat()},
+			{name: "server3", weight: randFloat()},
+		}
+
+		roundRobinServers = []server{
+			{name: "server1", weight: math.SmallestNonzeroFloat64},
+			{name: "server2", weight: math.SmallestNonzeroFloat64},
+			{name: "server3", weight: math.SmallestNonzeroFloat64},
+		}
+
+		rand.Shuffle(len(servers), func(i, j int) { servers[i], servers[j] = servers[j], servers[i] })
+	})
+
+	When("lb is created", func() {
+		It("its state is new", func() {
+			for _, l := range []*smoothWeightedRR{lb} {
+				validateEmptyLBState(l)
+			}
+		})
+	})
+
+	When("removing all items", func() {
+		It("should return an empty map", func() {
+			for _, l := range []*smoothWeightedRR{lb} {
+				addAllServers(l)
+				validateAllServersAdded(l)
+				l.RemoveAll()
+				validateEmptyLBState(l)
+			}
+		})
+	})
+
+	When("a single item is added", func() {
+		It("should fail on adding nil interface", func() {
```

Comment:
```suggestion
	When("a nil item is added", func() {
		It("should return an error", func() {
```

Reply:
Here you mean to split the When into two cases? because I have 4 more tests with an actual item, nil is only one test out of five in this case...

Reply:
OK, I think I got it, split when to each relevalt case

Reply:
yes - a `When` per case to clarify.